src/Raven.Server/Documents/Queries/Graph/RecursionQueryStep.cs

```diff
@@ -304,29 +306,42 @@ private void ProcessSingleResultRecursive(Match currentMatch, List<Match> matche
             _path.Push(new RecursionState { Src = startingPoint.Data, Match = currentMatch });
 
             int aliasBaseIndex = 0;
-
             Document cur = startingPoint;
+
+            //this function is needed for "documentation" purposes
+            bool AddMatchToResultsAndCheckIfNeedToStop(Document current)
+            {
+                if (AddMatchToResults(current))
+                    return true;
+
+                _path.Pop();
```

Comment:
Don't we need to clear the `_visited` here as well?

Reply:
Don't think so, because we are backtracking + clearing _visited only in the block that comes after ```if (top.Matches == null || top.Matches.Count == 0)```
(https://github.com/myarichuk/ravendb/blob/RavenDB-12263/src/Raven.Server/Documents/Queries/Graph/RecursionQueryStep.cs#L368)